Cancel creation of a cloud server

gitlaura opened the pull request:
- Keeping track of provisioners and installers in `cloudInterfaces` object in uproxy_core so that we can destroy them to cancel the install process
- Added cancel button to `installingOverlay`; button is disabled with `ui.cloudInstallCancelDisabled` if we are in the removing stage (if a server already existed) or the configuring ssh stage (which is the last stage and only lasts a few seconds)
- Added `cancelingOverlay` because canceling a cloud server can take up to a couple minutes if you try to cancel it at the very beginning. According to what I've read on Digital Ocean, you can't destroy a server while another action is in progress so we have to wait until the server is created. 

Fix for https://github.com/uProxy/uproxy/issues/2360
With minor changes to uproxy-lib: https://github.com/uProxy/uproxy-lib/pull/387

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/2367)
<!-- Reviewable:end -->

dborkan commented:
If you click to create a server, then immediately click cancel in the UI, it looks like we could call ```provisioner.stop(...)``` before the ```provisioner.start(...)``` call fulfills.  Is this a problem?  In the provisioner code itself does it wait for start to finish?

---

Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/generic_core/uproxy_core.ts, line 77 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2367#-KFekelaxieNvH_MJK70:-KFekelbT5itW2wvG9rd:1394324450) ([raw file](https://github.com/uproxy/uproxy/blob/34da308ab4bb16eb2451c709b361577ec4a19bdb/src/generic_core/uproxy_core.ts#L77)):*
Can you type this?  Then you don't need the comment on what the format is and the compiler will check that.  You can see an example at https://github.com/uProxy/uproxy/blob/master/src/generic_core/remote-user.ts#L61 (you might want to define a new interface just in this file for the value type)

---

*[src/generic_ui/polymer/cloud-install.ts, line 91 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2367#-KFfBAgpsFSoIZcistRX:-KFfBAgpsFSoIZcistRY:1956108622) ([raw file](https://github.com/uproxy/uproxy/blob/34da308ab4bb16eb2451c709b361577ec4a19bdb/src/generic_ui/polymer/cloud-install.ts#L91)):*
Should we have this in the ```open``` method as well?

---

*[src/generic_ui/scripts/ui.ts, line 312 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2367#-KFeyw1DuKfiGYAxb67N:-KFeyw1DuKfiGYAxb67O:898255267) ([raw file](https://github.com/uproxy/uproxy/blob/34da308ab4bb16eb2451c709b361577ec4a19bdb/src/generic_ui/scripts/ui.ts#L312)):*
Can you comment on why we can't cancel during the configuring ssh step?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2367#-:-KFfByPRxaJGwRuzkYfy:1499193543)*
<!-- Sent from Reviewable.io -->

gitlaura commented:
I have seen this scenario. Provisioner.stop receives an error that 'the server doesn't exist' and resolves successfully, then the installer and provisioner are destroyed (which is the same as in all other cases) and provisioner.start just fails silently. The cancellation process is successful in less than a second. 

---

Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/generic_core/uproxy_core.ts, line 77 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2367#-KFekelaxieNvH_MJK70:-KFj7gWhA6SbtPs8GMjT:1706342937) ([raw file](https://github.com/uproxy/uproxy/blob/34da308ab4bb16eb2451c709b361577ec4a19bdb/src/generic_core/uproxy_core.ts#L77)):*
Done. 

---

*[src/generic_ui/polymer/cloud-install.ts, line 91 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2367#-KFfBAgpsFSoIZcistRX:-KFirad_A8yNxoUNiczx:-1551040201) ([raw file](https://github.com/uproxy/uproxy/blob/34da308ab4bb16eb2451c709b361577ec4a19bdb/src/generic_ui/polymer/cloud-install.ts#L91)):*
Do you mean before the installOverlay.open() in loginTapped? I added it there too.

---

*[src/generic_ui/scripts/ui.ts, line 312 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2367#-KFeyw1DuKfiGYAxb67N:-KFitoiuWngfosyK5xon:-1714711089) ([raw file](https://github.com/uproxy/uproxy/blob/34da308ab4bb16eb2451c709b361577ec4a19bdb/src/generic_ui/scripts/ui.ts#L312)):*
The reason I am not allowing the user to cancel during the last step is mostly because it gets messy. Cloud may or may not be logged in and the user may or may not have accepted the invitation. This stage also usually only lasts a couple seconds. Added a comment. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2367#-:-KFiqB9o3zgDpuponKGw:-1532210994)*
<!-- Sent from Reviewable.io -->

dborkan commented:
OK - I'm a little confused as to why provisioner.start fails in that case but it sounds like what we want.  The race condition I want to avoid is where provisioner.start doesn't finish yet, then we call provisioner.stop and it returns success, but then provisioner.start also returns success and you have a server created that you are paying for (hopefully at least it would appear in the UI in that case)..

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/generic_ui/polymer/cloud-install.ts, line 91 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2367#-KFfBAgpsFSoIZcistRX:-KFoJELCFTkmmjGIGq8m:-1479679421) ([raw file](https://github.com/uproxy/uproxy/blob/34da308ab4bb16eb2451c709b361577ec4a19bdb/src/generic_ui/polymer/cloud-install.ts#L91)):*
I was thinking the open method in this file, line 14

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2367#-:-KFoJWoJEIhHkd58n0Qv:-318478099)*
<!-- Sent from Reviewable.io -->

gitlaura commented:
Just looked into it more. It fails because the provisioner tries to sendStatus during setupDigitalOcean_ but the messages cannot be delivered because the interface has been destroyed. But now that you say it, I think there is a very slim possibility that we could try to create the server after destroyServer succeeds. One solution is to disable the cancel button until after the 'creating server' phase since right now we have to wait for that stage to complete anyways before canceling. Thoughts? Technically the same thing will happen, I think the user could just be annoyed that they can't push the cancel button right away.

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2367#-:-KFoMpKZ9r3rTRcYW-KD:225874621)*
<!-- Sent from Reviewable.io -->

dborkan commented:
I think things are probably fine for now.  Maybe can we just enter a new GitHub issue to clean this up?  I could imagine that you might want to store the Promise returned by .start, then if you get a cancel call stop, wait for the .start Promise to fulfill/reject, then if fulfills (the race condition case) call .stop again?

I think even in the worst case now the user will at least see the newly created server on their screen so they can delete it.. 

Also this pull request suddenly got a lot more files in it, I'm guessing that some issue with merging?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2367#-:-KFoTc1qgUlamrynFUEL:-2086180022)*
<!-- Sent from Reviewable.io -->

gitlaura commented:
Okay sounds good. You're right that the user could just delete the server. 

And yikes! I thought I just merged the master branch in. Will fix now. 

---

Review status: 0 of 65 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2367#-:-KFoebF4Z4UNywAjCKb5:1448946355)*
<!-- Sent from Reviewable.io -->

gitlaura commented:
Review status: 0 of 65 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/generic_ui/polymer/cloud-install.ts, line 91 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2367#-KFfBAgpsFSoIZcistRX:-KFp7i6Zo7rt6BoOVUd4:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/34da308ab4bb16eb2451c709b361577ec4a19bdb/src/generic_ui/polymer/cloud-install.ts#L91)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2367)*
<!-- Sent from Reviewable.io -->
